File: frontend/src/features/copilot/types/conversation.ts
import type { Message } from '@ai-sdk/react';

import type { ConversationSummary } from '../../../contracts/generated/schemas';
import type { AnalysisResult } from './api';

export type { ConversationSummary };

export interface ConversationDetail {
  id: string;
  title: string;
  createdAt: number;
  updatedAt: number;
  messages: Message[];
  analysesById: Record<string, AnalysisResult>;
  messageAnalysisIds: Record<string, string>;
}

File: backend/app/models/conversation.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.models.analysis import AnalysisResult


class ConversationMessage(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    role: Literal["user", "assistant", "system"]
    content: str


class ConversationSummary(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    title: str
    created_at: int = Field(alias="createdAt")
    updated_at: int = Field(alias="updatedAt")


class ConversationDetail(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    title: str
    created_at: int = Field(alias="createdAt")
    updated_at: int = Field(alias="updatedAt")
    messages: list[ConversationMessage] = Field(default_factory=list)
    analyses_by_id: dict[str, AnalysisResult] = Field(
        default_factory=dict,
        alias="analysesById",
    )
    message_analysis_ids: dict[str, str] = Field(
        default_factory=dict,
        alias="messageAnalysisIds",
    )


# PUT body is the same shape as the detail response (full replace upsert).
ConversationUpsert = ConversationDetail
